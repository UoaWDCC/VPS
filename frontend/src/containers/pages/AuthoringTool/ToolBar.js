import React, { useState } from "react";
import {
  Box,
  Tooltip,
  Button,
  MenuList,
  Menu,
  MenuItem,
} from "@material-ui/core";
import ImageIcon from "@material-ui/icons/Image";
import TextFieldsIcon from "@material-ui/icons/TextFields";
import VolumeUpIcon from "@material-ui/icons/VolumeUp";
import styles from "../../../styling/ToolBar.module.scss";

const SidebarData = [
  {
    title: "Background Image",
    icon: <ImageIcon fontSize="medium" />,
    dropdown: {},
    onClick: null,
  },
  {
    title: "Text",
    icon: <TextFieldsIcon fontSize="medium" />,
    dropdown: {},
    onClick: null,
  },
  {
    title: "Audio",
    icon: <VolumeUpIcon fontSize="medium" />,
    dropdown: null,
    onClick: null,
  },
];

export default function ToolBar() {
  const [anchorEl, setAnchorEl] = useState(null);
  const handleDropdownClick = (e) => {
    setAnchorEl(e.currentTarget);
  };

  const handleDropdownClose = () => {
    setAnchorEl(null);
  };
  return (
    <>
      <MenuList className={styles.toolBar}>
        <Box display="flex" flexDirection="row">
          {SidebarData.map((val) => {
            const menuOnClick = val.dropdown
              ? handleDropdownClick
              : val.onClick;
            const open = val.title === anchorEl?.getAttribute("title");
            return (
              <>
                <Tooltip title={val.title} enterDelay={200} key={val.title}>
                  <Button
                    className={styles.menuItem}
                    key={val.title}
                    onClick={menuOnClick}
                  >
                    {val.icon}
                  </Button>
                </Tooltip>
                <Menu
                  id={val.title}
                  open={open}
                  anchorEl={anchorEl}
                  getContentAnchorEl={null}
                  onClose={handleDropdownClose}
                  anchorOrigin={{
                    vertical: "bottom",
                    horizontal: "left",
                  }}
                  transformOrigin={{
                    vertical: "top",
                    horizontal: "left",
                  }}
                >
                  <MenuItem onClick={() => {}}>Profile</MenuItem>
                </Menu>
              </>
            );
          })}
        </Box>
      </MenuList>
    </>
  );
}
