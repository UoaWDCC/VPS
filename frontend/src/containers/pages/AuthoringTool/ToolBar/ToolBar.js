import React, { useState, useContext } from "react";
import { Box, Tooltip, Button, MenuList, Menu } from "@material-ui/core";
import ArrowDropDownIcon from "@material-ui/icons/ArrowDropDown";
import toolBarData from "./ToolBarData";
import SceneContext from "../../../../context/SceneContext";
import styles from "../../../../styling/ToolBar.module.scss";

export default function ToolBar() {
  const { currentScene, setCurrentScene } = useContext(SceneContext);
  const [anchorEl, setAnchorEl] = useState(null);
  const [triggerElTitle, setTriggerElTitle] = useState(null);
  const handleDropdownClick = (e, title) => {
    setTriggerElTitle(title);
    setAnchorEl(e.currentTarget);
  };

  const handleDropdownClose = () => {
    setTriggerElTitle(null);
    setAnchorEl(null);
  };
  return (
    <>
      <MenuList className={styles.toolBar}>
        <Box display="flex" flexDirection="row">
          {toolBarData.map((tool) => {
            const menuOnClick = tool.dropdown
              ? (event) => {
                  handleDropdownClick(event, tool.title);
                }
              : () => tool.onClick(currentScene, setCurrentScene);
            const open = tool.title === triggerElTitle;
            return (
              <div key={tool.title}>
                <Tooltip title={tool.title} enterDelay={200} key={tool.title}>
                  <Button
                    className={styles.menuItem}
                    key={tool.title}
                    onClick={menuOnClick}
                  >
                    {tool.icon}
                    {tool.dropdown && <ArrowDropDownIcon fontSize="small" />}
                  </Button>
                </Tooltip>
                {tool.dropdown && (
                  <SubMenu
                    tool={tool}
                    open={open}
                    anchorEl={anchorEl}
                    handleDropdownClose={handleDropdownClose}
                  />
                )}
              </div>
            );
          })}
        </Box>
      </MenuList>
    </>
  );
}

// This is the tool dropdown
// It is only rendered when the tool component has a dropdown
const SubMenu = ({ tool, open, anchorEl, handleDropdownClose }) => {
  return (
    <Menu
      id={tool.title}
      open={open}
      anchorEl={anchorEl}
      getContentAnchorEl={null}
      onClose={handleDropdownClose}
      anchorOrigin={{
        vertical: "bottom",
        horizontal: "left",
      }}
      transformOrigin={{
        vertical: "top",
        horizontal: "left",
      }}
    >
      {tool.dropdown.map((dropdown, index) => {
        // eslint-disable-next-line react/no-array-index-key
        return <div key={index}>{dropdown.component}</div>;
      })}
    </Menu>
  );
};
