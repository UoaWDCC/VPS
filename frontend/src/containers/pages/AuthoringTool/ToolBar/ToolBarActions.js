import { v4 } from "uuid";

function addButton(currentScene, setCurrentScene) {
  const newButton = {
    type: "BUTTON",
    text: "Button",
    variant: "contained",
    colour: "white",
    nextScene: "",
    left: 0, // as percentage
    top: 0, // as percentage
    height: 6, // as percentage
    width: 20, // as percentage
    id: v4(),
  };

  const updatedComponents = currentScene.components;

  updatedComponents.push(newButton);

  setCurrentScene({
    ...currentScene,
    components: updatedComponents,
  });
}

function addText(currentScene, setCurrentScene) {
  const newText = {
    type: "TEXT",
    text: "default text",
    border: true,
    fontSize: 16, // pt
    color: "black",
    textAlign: "left",
    left: 0, // as percentage
    top: 0, // as percentage
    height: 10, // as percentage
    width: 20, // as percentage
    id: v4(),
  };

  const updatedComponents = currentScene.components;

  updatedComponents.push(newText);

  setCurrentScene({
    ...currentScene,
    components: updatedComponents,
  });
}

function addImage(currentScene, setCurrentScene, image) {
  const newImage = {
    type: "IMAGE",
    imageId: image._id,
    left: 0, // as percentage
    top: 0, // as percentage
    height: "auto", // as percentage
    width: "auto", // as percentage
    id: v4(),
  };

  const updatedComponents = currentScene.components;

  updatedComponents.push(newImage);

  setCurrentScene({
    ...currentScene,
    components: updatedComponents,
  });
}

function addFirebaseImage(currentScene, setCurrentScene, fileObject, url) {
  const newImage = {
    type: "FIREBASEIMAGE",
    fileObject,
    url,
    left: 0, // as percentage
    top: 0, // as percentage
    height: "auto", // as percentage
    width: "auto", // as percentage
    id: v4(),
  };

  const updatedComponents = currentScene.components;

  updatedComponents.push(newImage);

  setCurrentScene({
    ...currentScene,
    components: updatedComponents,
  });
}

function addFirebaseAudio(currentScene, setCurrentScene, fileObject, url) {
  const newAudio = {
    type: "FIREBASEAUDIO",
    name: fileObject.name,
    fileObject,
    url,
    loop: false,
    left: 0, // as percentage
    top: 0, // as percentage
    height: 10, // as percentage
    width: "auto", // icon ratio is fixed
    id: v4(),
  };

  const updatedComponents = currentScene.components;

  updatedComponents.push(newAudio);

  setCurrentScene({
    ...currentScene,
    components: updatedComponents,
  });
}

export { addButton, addText, addImage, addFirebaseImage, addFirebaseAudio };
