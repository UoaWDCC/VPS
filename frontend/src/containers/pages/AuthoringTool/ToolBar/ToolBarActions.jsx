import { v4 } from "uuid";

/**
 * @param {object} currentScene
 * @param {function} setCurrentScene
 */
function addComponent(component, currentScene, setCurrentScene) {
  const updatedComponents = currentScene.components;

  updatedComponents.push(component);

  setCurrentScene({
    ...currentScene,
    components: updatedComponents,
  });
}

/**
 * function to be put into ToolBarData when button is added
 */
function addButton(currentScene, setCurrentScene) {
  const newButton = {
    type: "BUTTON",
    text: "Button",
    variant: "contained",
    colour: "white",
    nextScene: "",
    left: 0, // as percentage
    top: 0, // as percentage
    height: 6, // as percentage
    width: 20, // as percentage
    id: v4(),
  };

  addComponent(newButton, currentScene, setCurrentScene);
}

/**
 * function to be put into ToolBarData when text is added
 */
function addText(currentScene, setCurrentScene) {
  const newText = {
    type: "TEXT",
    text: "default text",
    border: true,
    fontSize: 16, // pt
    color: "black",
    textAlign: "left",
    left: 0, // as percentage
    top: 0, // as percentage
    height: 10, // as percentage
    width: 20, // as percentage
    id: v4(),
  };

  addComponent(newText, currentScene, setCurrentScene);
}

function addSpeechText(currentScene, setCurrentScene) {
  const newSpeechText = {
    type: "SPEECHTEXT",
    text: "default speech text",
    border: true,
    fontSize: 16, // pt
    color: "black",
    textAlign: "left",
    left: 0, // as percentage
    top: 0, // as percentage
    height: 10, // as percentage
    width: 20, // as percentage
    arrowLocation: "top",
    id: v4(),
  };

  addComponent(newSpeechText, currentScene, setCurrentScene);
}

/**
 * function to be put into ToolBarData when image is added
 * @param {object} image
 */
function addImage(currentScene, setCurrentScene, image) {
  const newImage = {
    type: "IMAGE",
    imageId: image._id,
    left: 0, // as percentage
    top: 0, // as percentage
    height: "auto", // as percentage
    width: "auto", // as percentage
    id: v4(),
  };

  addComponent(newImage, currentScene, setCurrentScene);
}

/**
 * function to be put into ToolBarData when firebase image is added
 * @param {object} fileObject
 * @param {string} url
 */
function addFirebaseImage(currentScene, setCurrentScene, fileObject, url) {
  const newImage = {
    type: "FIREBASEIMAGE",
    fileObject,
    url,
    left: 0, // as percentage
    top: 0, // as percentage
    height: "auto", // as percentage
    width: "auto", // as percentage
    id: v4(),
  };

  addComponent(newImage, currentScene, setCurrentScene);
}

/**
 * call this when user adds firebase audio
 *
 * @param {object} fileObject
 * @param {string} url
 */
function addFirebaseAudio(currentScene, setCurrentScene, fileObject, url) {
  const newAudio = {
    type: "FIREBASEAUDIO",
    name: fileObject.name,
    fileObject,
    url,
    loop: false,
    left: 0, // as percentage
    top: 0, // as percentage
    height: 10, // as percentage
    width: 5, // as percentage
    id: v4(),
  };

  addComponent(newAudio, currentScene, setCurrentScene);
}

export {
  addButton,
  addFirebaseAudio,
  addFirebaseImage,
  addImage,
  addSpeechText,
  addText,
};
