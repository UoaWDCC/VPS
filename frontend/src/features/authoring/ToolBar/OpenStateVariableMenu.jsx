import { MenuItem } from "@material-ui/core";
import { useState } from "react";
import styles from "./ToolBar.module.scss";
import StateVariableMenu from "../../../components/StateVariables/StateVariableMenu";

/**
 * Used to open the State Variable Menu
 * @component
 */
const OpenStateVariableMenu = () => {
  const [show, setShow] = useState(false);

  return (
    <div>
      <MenuItem
        className={styles.menuItem}
        color="default"
        variant="contained"
        onClick={() => setShow(true)}
      >
        Open State Menu
      </MenuItem>
      <StateVariableMenu show={show} setShow={setShow} />
    </div>
  );
};

export default OpenStateVariableMenu;
