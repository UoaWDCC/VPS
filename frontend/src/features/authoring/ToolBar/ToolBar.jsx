import { useContext } from "react";

import { Box, Button, Menu, MenuList, Tooltip } from "@material-ui/core";
import ArrowDropDownIcon from "@material-ui/icons/ArrowDropDown";

import AuthoringToolContext from "context/AuthoringToolContext";
import SceneContext from "context/SceneContext";
import ToolbarContext from "context/ToolbarContext";
import styles from "./ToolBar.module.scss";
import toolBarData from "./ToolBarData";

/**
 * Toolbar with icons for adding scene components
 * @component
 */
export default function ToolBar() {
  const { currentScene, setCurrentScene } = useContext(SceneContext);
  const { setSelect } = useContext(AuthoringToolContext);
  const { anchorEl, triggerElTitle, handleDropdownClick, handleDropdownClose } =
    useContext(ToolbarContext);

  return (
    <>
      <MenuList className={styles.toolBar}>
        <Box
          display="flex"
          flexDirection="row"
          onClick={() => {
            setSelect(null);
          }}
        >
          {toolBarData.map((tool) => {
            const menuOnClick = tool.dropdown
              ? (event) => {
                  handleDropdownClick(event, tool.title);
                }
              : () => tool.onClick(currentScene, setCurrentScene);
            return (
              <div key={tool.title}>
                <Tooltip title={tool.title} enterDelay={200} key={tool.title}>
                  <Button
                    className={styles.menuItem}
                    key={tool.title}
                    onClick={menuOnClick}
                  >
                    {tool.icon} <p>Add {tool.title}</p>
                    {tool.dropdown && <ArrowDropDownIcon fontSize="small" />}
                  </Button>
                </Tooltip>
                {tool.dropdown && (
                  <SubMenu
                    tool={tool}
                    open={tool.title === triggerElTitle}
                    anchorEl={anchorEl}
                    handleDropdownClose={handleDropdownClose}
                  />
                )}
              </div>
            );
          })}
        </Box>
      </MenuList>
    </>
  );
}

// This is the tool dropdown
// It is only rendered when the tool component has a dropdown
const SubMenu = ({ tool, open, anchorEl, handleDropdownClose }) => {
  return (
    <Menu
      id={tool.title}
      open={open}
      anchorEl={anchorEl}
      getContentAnchorEl={null}
      onClose={handleDropdownClose}
      anchorOrigin={{
        vertical: "bottom",
        horizontal: "left",
      }}
      transformOrigin={{
        vertical: "top",
        horizontal: "left",
      }}
    >
      {tool.dropdown.map((dropdown, index) => {
        // eslint-disable-next-line react/no-array-index-key
        return <div key={index}>{dropdown.component}</div>;
      })}
    </Menu>
  );
};
