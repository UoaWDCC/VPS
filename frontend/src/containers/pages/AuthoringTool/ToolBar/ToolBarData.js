import React from "react";
import ImageIcon from "@material-ui/icons/Image";
import TextFieldsIcon from "@material-ui/icons/TextFields";
import ButtonIcon from "@material-ui/icons/AddCircle";
import VolumeUpIcon from "@material-ui/icons/VolumeUp";
import ChatBubbleIcon from "@material-ui/icons/ChatBubble";
import ChooseBackgroundSubMenu from "./Background/ChooseBackgroundSubMenu";
import { addButton, addText, addSpeechText } from "./ToolBarActions";
import UploadImage from "./Background/UploadImage";
import UploadAudio from "./Audio/UploadAudio";
import PlayersListSubMenu from "./Background/PlayersListSubMenu";

/**
 * This file contains the data for the add component icons to be added into the ToolBar
 */
const toolBarData = [
  {
    title: "Image",
    icon: <ImageIcon fontSize="medium" />,
    dropdown: [
      {
        component: <ChooseBackgroundSubMenu />,
      },
      {
        component: <UploadImage />,
      },
    ],
  },
  {
    title: "Text",
    icon: <TextFieldsIcon fontSize="medium" />,
    onClick: addText,
  },
  {
    title: "Button",
    icon: <ButtonIcon fontSize="medium" />,
    onClick: addButton,
  },
  {
    title: "Audio",
    icon: <VolumeUpIcon fontSize="medium" />,
    dropdown: [
      {
        component: <UploadAudio />,
      },
    ],
  },
  {
    title: "Speech text",
    icon: <ChatBubbleIcon fontSize="medium" />,
    onClick: addSpeechText,
  },
  {
    title: "Apply Tags",
    icon: <ImageIcon fontSize="medium" />,
    dropdown: [
      {
        component: <PlayersListSubMenu />,
      },
    ],
  },
];

export default toolBarData;
