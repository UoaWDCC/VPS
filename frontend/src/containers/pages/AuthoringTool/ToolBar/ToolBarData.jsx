import ButtonIcon from "@material-ui/icons/AddCircle";
import ChatBubbleIcon from "@material-ui/icons/ChatBubble";
import ImageIcon from "@material-ui/icons/Image";
import TextFieldsIcon from "@material-ui/icons/TextFields";
import VolumeUpIcon from "@material-ui/icons/VolumeUp";
import UploadAudio from "./Audio/UploadAudio";
import ChooseBackgroundSubMenu from "./Background/ChooseBackgroundSubMenu";
import UploadImage from "./Background/UploadImage";
import { addButton, addSpeechText, addText } from "./ToolBarActions";

/**
 * This file contains the data for the add component icons to be added into the ToolBar
 */
const toolBarData = [
  {
    title: "Image",
    icon: <ImageIcon fontSize="medium" />,
    dropdown: [
      {
        component: <ChooseBackgroundSubMenu />,
      },
      {
        component: <UploadImage />,
      },
    ],
  },
  {
    title: "Text",
    icon: <TextFieldsIcon fontSize="medium" />,
    onClick: addText,
  },
  {
    title: "Button",
    icon: <ButtonIcon fontSize="medium" />,
    onClick: addButton,
  },
  {
    title: "Audio",
    icon: <VolumeUpIcon fontSize="medium" />,
    dropdown: [
      {
        component: <UploadAudio />,
      },
    ],
  },
  {
    title: "Speech text",
    icon: <ChatBubbleIcon fontSize="medium" />,
    onClick: addSpeechText,
  },
];

export default toolBarData;
