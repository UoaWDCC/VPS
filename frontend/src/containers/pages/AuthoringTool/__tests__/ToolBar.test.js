import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import ToolBar from "../ToolBar/ToolBar";

test("Scenario Selection page snapshot test", () => {
  const component = renderer.create(
    <BrowserRouter>
      <ToolBar />
    </BrowserRouter>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
