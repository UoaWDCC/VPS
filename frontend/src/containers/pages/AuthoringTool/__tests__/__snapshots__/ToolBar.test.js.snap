// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scenario Selection page snapshot test 1`] = `
<div
  className="toolBar"
/>
`;
